File: src/lib/physics/caco3.ts
/**
 * §6 — CaCO₃ Geochemical Precipitation (Prong 2, non-ureolytic via Carbonic Anhydrase)
 * ===================================================================================
 * A GEOCHEMICAL precipitation model (per Lassin et al. 2018) — NOT a full biological MICP
 * model. Scope: simplified Ca–CO₂–H₂O chemistry around pH 8.5–10.5, ~25 °C.
 *
 * Cascade:
 *   pH + DIC → carbonate speciation (α0,α1,α2)
 *           → [CO₃²⁻]
 *   Ω = [Ca²⁺][CO₃²⁻]/Ksp ,  SI = log₁₀ Ω        (supersaturation)
 *   aqueous ions → ACC (fast, kinetic precursor) → calcite (ripening)   two-step
 *   calcite wt% → UCS (empirical power law)        biocement strength
 *   mol CaCO₃ → mol CO₂ captured                   climate co-benefit
 *
 * Carbonic anhydrase accelerates CO₂(aq)+H₂O → H⁺+HCO₃⁻, raising DIC (hence [CO₃²⁻] and Ω)
 * far faster than the un-catalyzed hydration — without ureolysis, so no ammonia.
 */

import { CACO3_CALIB, MOLAR_MASS, cval } from "./constants";

const pow10 = (x: number) => Math.pow(10, x);

export interface Speciation {
  /** α0 — fraction as dissolved CO₂(aq)/H₂CO₃*. */
  alpha0: number;
  /** α1 — fraction as bicarbonate HCO₃⁻. */
  alpha1: number;
  /** α2 — fraction as carbonate CO₃²⁻. */
  alpha2: number;
}

/**
 * Carbonate speciation fractions of total dissolved inorganic carbon at a given pH.
 *   α0 = [CO₂]/DIC, α1 = [HCO₃⁻]/DIC, α2 = [CO₃²⁻]/DIC
 */
export function carbonateSpeciation(pH: number): Speciation {
  const H = pow10(-pH);
  const Ka1 = pow10(-cval(CACO3_CALIB.pKa1));
  const Ka2 = pow10(-cval(CACO3_CALIB.pKa2));
  const denom = H * H + Ka1 * H + Ka1 * Ka2;
  return {
    alpha0: (H * H) / denom,
    alpha1: (Ka1 * H) / denom,
    alpha2: (Ka1 * Ka2) / denom,
  };
}

/** Saturation ratio Ω = [Ca²⁺][CO₃²⁻]/Ksp (concentrations in mol·L⁻¹). */
export function saturationRatio(
  caMolar: number,
  co3Molar: number,
  pKsp: number,
): number {
  const Ksp = pow10(-pKsp);
  return (caMolar * co3Molar) / Ksp;
}

/** Saturation index SI = log₁₀ Ω. SI>0 ⇒ precipitation thermodynamically favorable. */
export function saturationIndex(
  caMolar: number,
  co3Molar: number,
  pKsp: number,
): number {
  const omega = saturationRatio(caMolar, co3Molar, pKsp);
  return omega > 0 ? Math.log10(omega) : -Infinity;
}

/**
 * Normalized CA activity 0–1 from the fraction of maximal rate enhancement realized.
 *   activity = log₁₀(1 + f·enhancement) / log₁₀(1 + enhancement)
 * (log scaling because CA enhancement spans orders of magnitude).
 */
export function caActivityFraction(
  realizedEnhancementFraction: number,
): number {
  const E = cval(CACO3_CALIB.caRateEnhancement);
  const f = Math.max(0, Math.min(1, realizedEnhancementFraction));
  return Math.log10(1 + f * E) / Math.log10(1 + E);
}

export interface PrecipitationInputs {
  /** Initial dissolved Ca²⁺ [mM]. */
  calciumMillimolar: number;
  /** Maximum attainable DIC [mM] when CA fully active (CO₂ source capacity). */
  dicMaxMillimolar: number;
  /** Solution pH (clamped to the model's valid 8.5–10.5 window). */
  pH: number;
  /** Normalized CA activity 0–1 (see caActivityFraction). */
  caActivity: number;
  /** Sand bulk dry mass per litre of pore solution [g/L] (for wt% conversion). */
  sandMassPerLiter?: number;
  /** Simulated duration [h]. */
  hours?: number;
  /** Integration step [h]. */
  dt?: number;
}

export interface PrecipitationStep {
  time: number; // h
  caMolar: number; // mol/L dissolved Ca²⁺
  co3Molar: number; // mol/L carbonate ion
  accMolar: number; // mol/L amorphous CaCO₃
  vateriteMolar: number; // mol/L metastable vaterite
  calciteMolar: number; // mol/L crystalline calcite
  SIcalcite: number; // saturation index wrt calcite
  omegaACC: number; // saturation ratio wrt ACC
}

export interface PrecipitationResult {
  series: PrecipitationStep[];
  /** Final calcite [mol/L]. */
  calciteMolar: number;
  /** Final metastable vaterite [mol/L]. */
  vateriteMolar: number;
  /** Final calcite content [wt% of sand]. */
  calciteWtPercent: number;
  /** Final vaterite content [wt% of sand]. */
  vateriteWtPercent: number;
  /** Total crystalline + amorphous carbonate content [wt% of sand]. */
  carbonateWtPercent: number;
  /** Strength-weighted "effective calcite" wt% that drives UCS (vaterite counts at a reduced factor). */
  loadBearingWtPercent: number;
  /** Vaterite as a fraction of total crystalline carbonate (0–1) — the polymorph purity readout. */
  vateriteFraction: number;
  /** Unconfined compressive strength [kPa]. */
  ucsKpa: number;
  /** CO₂ sequestered [g per litre solution]. */
  co2SequesteredGPerL: number;
}

/**
 * Three-phase ACC → vaterite → calcite precipitation kinetics, integrated explicitly.
 * Non-ureolytic carbonic-anhydrase MICP is dominated by CO₂ hydration and typically crystallises
 * a large fraction of METASTABLE VATERITE (Research Table, Prong 2) that only slowly recrystallises
 * to calcite (solution-mediated Ostwald ripening). Vaterite is softer and more soluble, so the crust
 * is weaker while vaterite-rich and stiffens as it converts. CA activity raises the quasi-equilibrium
 * DIC, hence [CO₃²⁻] and Ω.
 */
export function simulatePrecipitation(
  inp: PrecipitationInputs,
): PrecipitationResult {
  const hours = inp.hours ?? 48;
  const dt = inp.dt ?? 0.25;
  const steps = Math.round(hours / dt);
  const sandMass = inp.sandMassPerLiter ?? 1500; // ~ bulk sand in 1 L pore volume

  const pKspCal = cval(CACO3_CALIB.pKspCalcite);
  const pKspACC = cval(CACO3_CALIB.pKspACC);
  const kPrecip = cval(CACO3_CALIB.kPrecip);
  const kRipen = cval(CACO3_CALIB.kAccToCalcite);
  const fVaterite = cval(CACO3_CALIB.vateriteFraction); // ACC ripens mostly to vaterite in CA-MICP
  const kVatToCal = cval(CACO3_CALIB.kVateriteToCalcite); // slow vaterite → calcite recrystallisation
  const vatStrength = cval(CACO3_CALIB.vateriteStrengthFactor); // vaterite load-bearing vs calcite

  const pH = Math.max(8.5, Math.min(10.5, inp.pH));
  const { alpha2 } = carbonateSpeciation(pH);

  // Quasi-equilibrium DIC scales with CA activity (more enzyme ⇒ more bicarbonate captured).
  const dicTargetM =
    (inp.dicMaxMillimolar * Math.max(0, Math.min(1, inp.caActivity))) / 1000;
  const kDic = 0.6; // DIC relaxation toward target [h⁻¹] (fast CO₂ hydration once CA present)

  let ca = inp.calciumMillimolar / 1000; // mol/L
  let dic = 0;
  let acc = 0;
  let vaterite = 0;
  let calcite = 0;

  const series: PrecipitationStep[] = [];

  for (let i = 0; i <= steps; i++) {
    const co3 = alpha2 * dic;
    const omegaCal = saturationRatio(ca, co3, pKspCal);
    const omegaACC = saturationRatio(ca, co3, pKspACC);
    const SIcalcite = omegaCal > 0 ? Math.log10(omegaCal) : -Infinity;

    series.push({
      time: i * dt,
      caMolar: ca,
      co3Molar: co3,
      accMolar: acc,
      vateriteMolar: vaterite,
      calciteMolar: calcite,
      SIcalcite,
      omegaACC,
    });

    // CA-driven DIC supply (first-order relaxation to the activity-set target).
    const dDic = kDic * (dicTargetM - dic);

    // ACC precipitation when supersaturated wrt the amorphous phase (TST: rate ∝ Ω−1).
    const accDrive = Math.max(0, omegaACC - 1);
    const rAccPrecip = Math.min(kPrecip * accDrive, ca / dt, dic / dt); // can't exceed available ions
    // ACC ripens (first order); in CA-MICP it splits into mostly vaterite + some direct calcite.
    const rRipen = kRipen * acc;
    // Metastable vaterite slowly recrystallises to calcite (solution-mediated, first order).
    const rVatToCal = kVatToCal * vaterite;

    ca = Math.max(0, ca - rAccPrecip * dt);
    dic = Math.max(0, dic + (dDic - rAccPrecip) * dt);
    acc = Math.max(0, acc + (rAccPrecip - rRipen) * dt);
    vaterite = Math.max(0, vaterite + (rRipen * fVaterite - rVatToCal) * dt);
    calcite = Math.max(
      0,
      calcite + (rRipen * (1 - fVaterite) + rVatToCal) * dt,
    );
  }

  const gPerL = (mol: number) => mol * MOLAR_MASS.CaCO3;
  const calciteWtPercent = (gPerL(calcite) / sandMass) * 100;
  const vateriteWtPercent = (gPerL(vaterite) / sandMass) * 100;
  const carbonateWtPercent = (gPerL(calcite + vaterite + acc) / sandMass) * 100;
  // Vaterite bears load at a reduced factor until it converts — this is what drives UCS.
  const loadBearingWtPercent =
    calciteWtPercent + vatStrength * vateriteWtPercent;
  const crystalline = calcite + vaterite;
  const vateriteFraction = crystalline > 0 ? vaterite / crystalline : 0;
  const ucsKpa = calciteToUCS(loadBearingWtPercent);
  const co2SequesteredGPerL = (calcite + vaterite + acc) * MOLAR_MASS.CO2; // 1 mol CaCO₃ fixes 1 mol CO₂

  return {
    series,
    calciteMolar: calcite,
    vateriteMolar: vaterite,
    calciteWtPercent,
    vateriteWtPercent,
    carbonateWtPercent,
    loadBearingWtPercent,
    vateriteFraction,
    ucsKpa,
    co2SequesteredGPerL,
  };
}

/** Empirical biocement strength curve: UCS = kUcs · (load-bearing carbonate wt%)^nUcs  [kPa]. */
export function calciteToUCS(calciteWtPercent: number): number {
  if (calciteWtPercent <= 0) return 0;
  return (
    cval(CACO3_CALIB.kUcs) * Math.pow(calciteWtPercent, cval(CACO3_CALIB.nUcs))
  );
}

/** CO₂ captured per mole of CaCO₃ precipitated (1:1 stoichiometry) → mass [g]. */
export const co2Sequestered = (molCaCO3: number): number =>
  Math.max(0, molCaCO3) * MOLAR_MASS.CO2;

File: python_models/caco3.py
"""
CaCO3 precipitation and biocement strength - NYUAD iGEM 2026 Dunelock toolkit
============================================================================
Reproduces the "CaCO3 Precipitation" module: a geochemical Ca-CO2-H2O precipitation
model with an explicit ACC -> vaterite -> calcite polymorph cascade (non-ureolytic
carbonic-anhydrase MICP), and the calcite-to-strength (UCS) power law.

Exact port of src/lib/physics/caco3.ts.
  carbonate speciation alpha2 at pH (Bjerrum), Omega = [Ca][CO3]/Ksp, SI = log10 Omega
  ACC precipitates when Omega_ACC > 1 (TST rate ~ Omega-1), ripens (k_r) mostly to
  vaterite (fraction f_v) which slowly recrystallises to calcite (k_vc),
  UCS = kUcs * (load-bearing wt%)^nUcs.

Constants from CACO3_CALIB in src/lib/physics/constants.ts (pKa1 6.35, pKa2 10.33,
pKspCalcite 8.48, pKspACC 6.4, kPrecip 0.12, kAccToCalcite 0.05, vateriteFraction 0.6,
kVateriteToCalcite 0.02, vateriteStrengthFactor 0.55, kUcs 31.6, nUcs 1.5).
Run:  python caco3.py
"""

import numpy as np
import matplotlib.pyplot as plt

ORANGE, TEAL, ROSE, MAROON, ASH = "#D6884A", "#8FB3AC", "#C28A7C", "#6E1E18", "#8A7E75"
plt.rcParams.update({
    "figure.figsize": (7.2, 4.3), "font.size": 11, "figure.dpi": 130,
    "axes.spines.top": False, "axes.spines.right": False, "axes.edgecolor": ASH,
    "axes.grid": True, "grid.color": "#E7D8C4", "grid.linewidth": 0.8,
    "axes.titleweight": "bold", "figure.facecolor": "#FBF7F0", "axes.facecolor": "#FBF7F0",
})

pKa1, pKa2 = 6.35, 10.33
pKspCal, pKspACC = 8.48, 6.4
kPrecip, kRipen = 0.12, 0.05
fVaterite, kVatToCal, vatStrength = 0.6, 0.02, 0.55
kUcs, nUcs = 31.6, 1.5
M_CaCO3 = 100.0869


def alpha2(pH):
    H = 10.0 ** (-pH)
    Ka1, Ka2 = 10.0 ** (-pKa1), 10.0 ** (-pKa2)
    return Ka1 * Ka2 / (H * H + Ka1 * H + Ka1 * Ka2)


def sat_ratio(ca_M, co3_M, pKsp):
    return ca_M * co3_M / (10.0 ** (-pKsp))


def simulate(ca_mM=20.0, dic_max_mM=25.0, pH=9.5, ca_activity=0.9,
             hours=48.0, dt=0.25, sand_g_per_L=1500.0):
    steps = int(round(hours / dt))
    a2 = alpha2(np.clip(pH, 8.5, 10.5))
    dic_target = dic_max_mM * np.clip(ca_activity, 0, 1) / 1000.0
    kDic = 0.6
    ca, dic, acc, vaterite, calcite = ca_mM / 1000.0, 0.0, 0.0, 0.0, 0.0
    t = np.zeros(steps + 1)
    cal_series, vat_series, si_series = (np.zeros(steps + 1) for _ in range(3))
    for i in range(steps + 1):
        co3 = a2 * dic
        omegaCal = sat_ratio(ca, co3, pKspCal)
        omegaACC = sat_ratio(ca, co3, pKspACC)
        t[i] = i * dt
        cal_series[i], vat_series[i] = calcite, vaterite
        si_series[i] = np.log10(omegaCal) if omegaCal > 0 else -np.inf
        dDic = kDic * (dic_target - dic)
        accDrive = max(0.0, omegaACC - 1)
        rAcc = min(kPrecip * accDrive, ca / dt, dic / dt)
        rRipen = kRipen * acc
        rVat = kVatToCal * vaterite
        ca = max(0.0, ca - rAcc * dt)
        dic = max(0.0, dic + (dDic - rAcc) * dt)
        acc = max(0.0, acc + (rAcc - rRipen) * dt)
        vaterite = max(0.0, vaterite + (rRipen * fVaterite - rVat) * dt)
        calcite = max(0.0, calcite + (rRipen * (1 - fVaterite) + rVat) * dt)
    cal_wt = cal_series * M_CaCO3 / sand_g_per_L * 100
    vat_wt = vat_series * M_CaCO3 / sand_g_per_L * 100
    load_wt = cal_wt + vatStrength * vat_wt
    return t, cal_series, vat_series, si_series, load_wt


def calcite_to_ucs(wt_percent):
    wt = np.maximum(0.0, wt_percent)
    return kUcs * np.power(wt, nUcs)


def figures():
    figs = []

    # 1) Polymorph cascade over time.
    t, cal, vat, si, load = simulate()
    fig1, ax1 = plt.subplots()
    ax1.plot(t, vat * 1000, color=ROSE, lw=2.2, label="vaterite (metastable)")
    ax1.plot(t, cal * 1000, color=MAROON, lw=2.4, label="calcite (durable)")
    ax1.set_xlabel("time (h)")
    ax1.set_ylabel("carbonate  (mmol/L)")
    ax1.set_title("ACC -> vaterite -> calcite polymorph cascade")
    ax1.legend(frameon=False)
    fig1.tight_layout()
    figs.append((fig1, "caco3-1.png"))

    # 2) Biocement strength (UCS) power law vs load-bearing carbonate wt%.
    wt = np.linspace(0, 12, 200)
    fig2, ax2 = plt.subplots()
    ax2.plot(wt, calcite_to_ucs(wt) / 1000.0, color=ORANGE, lw=2.4)
    ax2.set_xlabel("load-bearing carbonate  (wt%)")
    ax2.set_ylabel("UCS  (MPa)")
    ax2.set_title(r"Biocement strength  UCS = $k\,w^{\,n}$  (n = 1.5)")
    fig2.tight_layout()
    figs.append((fig2, "caco3-2.png"))
    return figs


if __name__ == "__main__":
    for fig, name in figures():
        fig.savefig(name, bbox_inches="tight")
        print("wrote", name)
